File: sandboxes/sdk/go/toolbox.go
package gocracker

import (
	"bufio"
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/url"
	"strings"
	"time"
)

// ToolboxClient talks to the in-guest toolbox agent over a per-sandbox
// UDS + Firecracker-style CONNECT handshake. Stateless — every call
// dials a fresh socket to match the agent's Connection: close
// response shape. Thread-safe (no shared per-client state).
type ToolboxClient struct {
	UDSPath     string
	Port        uint32        // default 10023
	DialTimeout time.Duration // default 5s
}

// TBChannel values mirror internal/toolbox/agent.Channel*.
const (
	TBChannelStdin  = 0
	TBChannelStdout = 1
	TBChannelStderr = 2
	TBChannelExit   = 3
	TBChannelSignal = 4
)

// NewToolboxClient builds a ToolboxClient with default port + timeout.
func NewToolboxClient(udsPath string) *ToolboxClient {
	return &ToolboxClient{UDSPath: udsPath, Port: 10023, DialTimeout: 5 * time.Second}
}

// ToolboxError is returned by every toolbox-agent call.
type ToolboxError struct{ Message string }

func (e *ToolboxError) Error() string { return e.Message }

// ExecResult captures stdout/stderr/exit from a blocking Exec.
type ExecResult struct {
	ExitCode int32
	Stdout   []byte
	Stderr   []byte
}

// ExecOptions extends Exec call with env / workdir / stdin.
type ExecOptions struct {
	Env     []string
	WorkDir string
	Stdin   []byte
	Timeout time.Duration
}

// Exec runs a command in the guest and collects all stdout/stderr/exit.
// Blocking; for streaming use ExecStream.
func (t *ToolboxClient) Exec(ctx context.Context, cmd []string, opts ...ExecOptions) (*ExecResult, error) {
	o := ExecOptions{Timeout: 30 * time.Second}
	if len(opts) > 0 {
		o = opts[0]
		if o.Timeout == 0 {
			o.Timeout = 30 * time.Second
		}
	}
	result := &ExecResult{ExitCode: -1}
	frames, err := t.ExecStream(ctx, cmd, o)
	if err != nil {
		return nil, err
	}
	for f := range frames {
		switch f.Channel {
		case TBChannelStdout:
			result.Stdout = append(result.Stdout, f.Payload...)
		case TBChannelStderr:
			result.Stderr = append(result.Stderr, f.Payload...)
		case TBChannelExit:
			if len(f.Payload) >= 4 {
				result.ExitCode = int32(binary.BigEndian.Uint32(f.Payload[:4]))
			}
		case TBChannelSignal:
			// Control frame — ignore for Exec aggregation.
		}
	}
	return result, nil
}

// Frame is one message from ExecStream.
type Frame struct {
	Channel byte
	Payload []byte
}

// ExecStream returns a channel of Frames that closes when the exec
// finishes. Callers drain it to implement line-by-line streaming or
// interactive I/O. Errors during the setup phase return immediately;
// errors mid-stream close the channel.
func (t *ToolboxClient) ExecStream(ctx context.Context, cmd []string, opts ExecOptions) (<-chan Frame, error) {
	if opts.Timeout == 0 {
		opts.Timeout = 30 * time.Second
	}
	body := map[string]any{"cmd": cmd}
	if len(opts.Env) > 0 {
		body["env"] = opts.Env
	}
	if opts.WorkDir != "" {
		body["workdir"] = opts.WorkDir
	}
	bodyBytes, _ := json.Marshal(body)

	sock, err := t.dialConnect(ctx)
	if err != nil {
		return nil, err
	}
	_ = sock.SetDeadline(time.Now().Add(opts.Timeout))

	req := fmt.Sprintf(
		"POST /exec HTTP/1.0\r\nHost: x\r\nContent-Length: %d\r\nContent-Type: application/json\r\nConnection: close\r\n\r\n",
		len(bodyBytes),
	)
	if _, err := sock.Write([]byte(req)); err != nil {
		sock.Close()
		return nil, &ToolboxError{Message: "exec: write headers: " + err.Error()}
	}
	if _, err := sock.Write(bodyBytes); err != nil {
		sock.Close()
		return nil, &ToolboxError{Message: "exec: write body: " + err.Error()}
	}
	if len(opts.Stdin) > 0 {
		if err := writeFrame(sock, TBChannelStdin, opts.Stdin); err != nil {
			sock.Close()
			return nil, err
		}
		if err := writeFrame(sock, TBChannelStdin, nil); err != nil {
			sock.Close()
			return nil, err
		}
	}

	br := bufio.NewReader(sock)
	if _, err := br.ReadString('\n'); err != nil {
		sock.Close()
		return nil, &ToolboxError{Message: "exec: status: " + err.Error()}
	}
	for {
		line, err := br.ReadString('\n')
		if err != nil {
			sock.Close()
			return nil, &ToolboxError{Message: "exec: headers: " + err.Error()}
		}
		if line == "\r\n" || line == "\n" {
			break
		}
	}

	out := make(chan Frame, 4)
	go func() {
		defer close(out)
		defer sock.Close()
		for {
			var hdr [5]byte
			if _, err := io.ReadFull(br, hdr[:]); err != nil {
				return
			}
			channel := hdr[0]
			n := binary.BigEndian.Uint32(hdr[1:5])
			var payload []byte
			if n > 0 {
				payload = make([]byte, n)
				if _, err := io.ReadFull(br, payload); err != nil {
					return
				}
			}
			select {
			case out <- Frame{Channel: channel, Payload: payload}:
			case <-ctx.Done():
				return
			}
			if channel == TBChannelExit {
				return
			}
		}
	}()
	return out, nil
}

// ---- Files ------------------------------------------------------

// FileEntry is one entry from ListFiles.
type FileEntry struct {
	Name  string `json:"name"`
	Path  string `json:"path"`
	Kind  string `json:"kind"`  // "file" | "dir"
	Size  int64  `json:"size"`
}

// IsDir is a convenience wrapper over Kind == "dir".
func (e FileEntry) IsDir() bool { return e.Kind == "dir" }

// ListFiles returns the entries under the guest path.
func (t *ToolboxClient) ListFiles(ctx context.Context, path string) ([]FileEntry, error) {
	status, _, body, err := t.httpReq(ctx, "GET", "/files?path="+url.QueryEscape(path), nil, "")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("list_files: status=%d", status)}
	}
	var resp struct {
		Entries []FileEntry `json:"entries"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, &ToolboxError{Message: "list_files: decode: " + err.Error()}
	}
	return resp.Entries, nil
}

// Download returns the raw bytes of a guest file.
func (t *ToolboxClient) Download(ctx context.Context, path string) ([]byte, error) {
	status, _, body, err := t.httpReq(ctx, "GET", "/files/download?path="+url.QueryEscape(path), nil, "")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("download: status=%d", status)}
	}
	return body, nil
}

// Upload writes data to a guest file.
func (t *ToolboxClient) Upload(ctx context.Context, path string, data []byte) error {
	status, _, _, err := t.httpReq(ctx, "POST", "/files/upload?path="+url.QueryEscape(path), data, "application/octet-stream")
	if err != nil {
		return err
	}
	if status != 200 && status != 201 {
		return &ToolboxError{Message: fmt.Sprintf("upload: status=%d", status)}
	}
	return nil
}

// DeleteFile removes a guest file.
func (t *ToolboxClient) DeleteFile(ctx context.Context, path string) error {
	status, _, _, err := t.httpReq(ctx, "DELETE", "/files?path="+url.QueryEscape(path), nil, "")
	if err != nil {
		return err
	}
	if status != 200 && status != 204 {
		return &ToolboxError{Message: fmt.Sprintf("delete_file: status=%d", status)}
	}
	return nil
}

// Mkdir creates a directory. parents=true is mkdir -p.
func (t *ToolboxClient) Mkdir(ctx context.Context, path string, parents bool) error {
	body, _ := json.Marshal(map[string]any{"path": path, "all": parents})
	status, _, _, err := t.httpReq(ctx, "POST", "/files/mkdir", body, "application/json")
	if err != nil {
		return err
	}
	if status != 200 {
		return &ToolboxError{Message: fmt.Sprintf("mkdir: status=%d", status)}
	}
	return nil
}

// Rename moves a guest file or directory.
func (t *ToolboxClient) Rename(ctx context.Context, src, dst string) error {
	body, _ := json.Marshal(map[string]any{"old_path": src, "new_path": dst})
	status, _, _, err := t.httpReq(ctx, "POST", "/files/rename", body, "application/json")
	if err != nil {
		return err
	}
	if status != 200 {
		return &ToolboxError{Message: fmt.Sprintf("rename: status=%d", status)}
	}
	return nil
}

// Chmod changes a guest file's mode bits.
func (t *ToolboxClient) Chmod(ctx context.Context, path string, mode uint32) error {
	body, _ := json.Marshal(map[string]any{"path": path, "mode": mode})
	status, _, _, err := t.httpReq(ctx, "POST", "/files/chmod", body, "application/json")
	if err != nil {
		return err
	}
	if status != 200 {
		return &ToolboxError{Message: fmt.Sprintf("chmod: status=%d", status)}
	}
	return nil
}

// ---- Git --------------------------------------------------------

// GitClone clones repository into directory, optionally checking out ref.
func (t *ToolboxClient) GitClone(ctx context.Context, repository, directory, ref string) (map[string]any, error) {
	b := map[string]any{"repository": repository, "directory": directory}
	if ref != "" {
		b["ref"] = ref
	}
	body, _ := json.Marshal(b)
	status, _, respBody, err := t.httpReq(ctx, "POST", "/git/clone", body, "application/json")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("git_clone: status=%d body=%s", status, string(respBody))}
	}
	var out map[string]any
	_ = json.Unmarshal(respBody, &out)
	return out, nil
}

// GitStatus runs `git -C directory status --porcelain`.
func (t *ToolboxClient) GitStatus(ctx context.Context, directory string) (map[string]any, error) {
	body, _ := json.Marshal(map[string]any{"directory": directory})
	status, _, respBody, err := t.httpReq(ctx, "POST", "/git/status", body, "application/json")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("git_status: status=%d", status)}
	}
	var out map[string]any
	_ = json.Unmarshal(respBody, &out)
	return out, nil
}

// ---- Secrets ----------------------------------------------------

// SetSecret stores a name=value pair in the per-sandbox in-memory store.
func (t *ToolboxClient) SetSecret(ctx context.Context, name, value string) error {
	body, _ := json.Marshal(map[string]any{"name": name, "value": value})
	status, _, _, err := t.httpReq(ctx, "POST", "/secrets", body, "application/json")
	if err != nil {
		return err
	}
	if status != 200 && status != 201 {
		return &ToolboxError{Message: fmt.Sprintf("set_secret: status=%d", status)}
	}
	return nil
}

// ListSecrets returns the names of stored secrets (values never leave
// the sandbox).
func (t *ToolboxClient) ListSecrets(ctx context.Context) ([]string, error) {
	status, _, body, err := t.httpReq(ctx, "GET", "/secrets", nil, "")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("list_secrets: status=%d", status)}
	}
	var resp struct {
		Secrets []string `json:"secrets"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, &ToolboxError{Message: "list_secrets: decode: " + err.Error()}
	}
	return resp.Secrets, nil
}

// DeleteSecret removes a stored secret.
func (t *ToolboxClient) DeleteSecret(ctx context.Context, name string) error {
	status, _, _, err := t.httpReq(ctx, "DELETE", "/secrets/"+name, nil, "")
	if err != nil {
		return err
	}
	if status != 200 && status != 204 {
		return &ToolboxError{Message: fmt.Sprintf("delete_secret: status=%d", status)}
	}
	return nil
}

// Health calls the agent's /healthz.
func (t *ToolboxClient) Health(ctx context.Context) (map[string]any, error) {
	status, _, body, err := t.httpReq(ctx, "GET", "/healthz", nil, "")
	if err != nil {
		return nil, err
	}
	if status != 200 {
		return nil, &ToolboxError{Message: fmt.Sprintf("health: status=%d", status)}
	}
	var out map[string]any
	_ = json.Unmarshal(body, &out)
	return out, nil
}

// ---- HTTP over UDS ---------------------------------------------

func (t *ToolboxClient) dialConnect(ctx context.Context) (net.Conn, error) {
	dialer := &net.Dialer{Timeout: t.DialTimeout}
	sock, err := dialer.DialContext(ctx, "unix", t.UDSPath)
	if err != nil {
		return nil, &ToolboxError{Message: "dial " + t.UDSPath + ": " + err.Error()}
	}
	port := t.Port
	if port == 0 {
		port = 10023
	}
	if _, err := sock.Write([]byte(fmt.Sprintf("CONNECT %d\n", port))); err != nil {
		sock.Close()
		return nil, &ToolboxError{Message: "CONNECT write: " + err.Error()}
	}
	br := bufio.NewReader(sock)
	line, err := br.ReadString('\n')
	if err != nil {
		sock.Close()
		return nil, &ToolboxError{Message: "CONNECT read: " + err.Error()}
	}
	if !strings.HasPrefix(line, "OK") {
		sock.Close()
		return nil, &ToolboxError{Message: "CONNECT rejected: " + strings.TrimSpace(line)}
	}
	// Wrap the socket to expose the already-buffered bytes from br
	// on subsequent reads (usually empty; defense-in-depth).
	return &prefixedConn{Conn: sock, br: br}, nil
}

// prefixedConn reads from the bufio.Reader (which may hold bytes
// that arrived after the OK\n handshake) before falling through
// to the raw conn.
type prefixedConn struct {
	net.Conn
	br *bufio.Reader
}

func (p *prefixedConn) Read(b []byte) (int, error) { return p.br.Read(b) }

func (t *ToolboxClient) httpReq(ctx context.Context, method, path string, body []byte, contentType string) (int, map[string]string, []byte, error) {
	sock, err := t.dialConnect(ctx)
	if err != nil {
		return 0, nil, nil, err
	}
	defer sock.Close()
	_ = sock.SetDeadline(time.Now().Add(10 * time.Second))

	hdrs := fmt.Sprintf("%s %s HTTP/1.0\r\nHost: x\r\nConnection: close\r\n", method, path)
	if body != nil {
		hdrs += fmt.Sprintf("Content-Length: %d\r\nContent-Type: %s\r\n", len(body), contentType)
	}
	hdrs += "\r\n"
	if _, err := sock.Write([]byte(hdrs)); err != nil {
		return 0, nil, nil, &ToolboxError{Message: "write headers: " + err.Error()}
	}
	if body != nil {
		if _, err := sock.Write(body); err != nil {
			return 0, nil, nil, &ToolboxError{Message: "write body: " + err.Error()}
		}
	}

	br := bufio.NewReader(sock)
	statusLine, err := br.ReadString('\n')
	if err != nil {
		return 0, nil, nil, &ToolboxError{Message: "read status: " + err.Error()}
	}
	parts := strings.SplitN(statusLine, " ", 3)
	if len(parts) < 2 {
		return 0, nil, nil, &ToolboxError{Message: "malformed status: " + strings.TrimSpace(statusLine)}
	}
	var status int
	fmt.Sscanf(parts[1], "%d", &status)

	headers := map[string]string{}
	for {
		line, err := br.ReadString('\n')
		if err != nil {
			return 0, nil, nil, &ToolboxError{Message: "read headers: " + err.Error()}
		}
		if line == "\r\n" || line == "\n" {
			break
		}
		if idx := strings.Index(line, ":"); idx > 0 {
			headers[strings.ToLower(strings.TrimSpace(line[:idx]))] = strings.TrimSpace(line[idx+1:])
		}
	}

	// Bounded body read: skip for 204/205/304, use Content-Length
	// when present, otherwise read to EOF.
	var bodyOut []byte
	if status != 204 && status != 205 && status != 304 {
		if cl, ok := headers["content-length"]; ok && cl != "" {
			var n int
			fmt.Sscanf(cl, "%d", &n)
			if n > 0 {
				bodyOut = make([]byte, n)
				_, err := io.ReadFull(br, bodyOut)
				if err != nil {
					return 0, nil, nil, &ToolboxError{Message: "read body: " + err.Error()}
				}
			}
		} else {
			bodyOut, _ = io.ReadAll(br)
		}
	}
	return status, headers, bodyOut, nil
}

func writeFrame(w io.Writer, channel byte, payload []byte) error {
	hdr := [5]byte{channel, 0, 0, 0, 0}
	binary.BigEndian.PutUint32(hdr[1:5], uint32(len(payload)))
	if _, err := w.Write(hdr[:]); err != nil {
		return &ToolboxError{Message: "write frame header: " + err.Error()}
	}
	if len(payload) > 0 {
		if _, err := w.Write(payload); err != nil {
			return &ToolboxError{Message: "write frame payload: " + err.Error()}
		}
	}
	return nil
}

File: sandboxes/sdk/python/gocracker/toolbox.py
"""Toolbox agent client over UDS + Firecracker-style CONNECT.

Dials the sandbox's host-side UDS, sends "CONNECT <port>\\n",
validates the "OK\\n" response, then speaks the toolbox HTTP
protocol on the bridged stream. Three operation groups:

  health():                          cheap liveness
  exec(cmd, ...):                    /exec framed binary protocol
  upload(path, data), download(path): /files JSON
"""
from __future__ import annotations

import io
import json
import socket
import struct
from dataclasses import dataclass, field
from typing import Dict, Iterator, List, Optional, Tuple


TOOLBOX_VSOCK_PORT = 10023

# Frame protocol (matches internal/toolbox/agent/frame.go):
#   [1 byte channel][4 byte length BE][payload]
CHANNEL_STDIN = 0
CHANNEL_STDOUT = 1
CHANNEL_STDERR = 2
CHANNEL_EXIT = 3
CHANNEL_SIGNAL = 4


class ToolboxError(Exception):
    """Base class for toolbox-agent client errors."""


class ToolboxTimeout(ToolboxError):
    """Raised when a UDS dial / read exceeds the deadline."""


@dataclass
class ExecResult:
    exit_code: int
    stdout: bytes
    stderr: bytes

    @property
    def stdout_text(self) -> str:
        return self.stdout.decode("utf-8", errors="replace")

    @property
    def stderr_text(self) -> str:
        return self.stderr.decode("utf-8", errors="replace")


@dataclass
class FileEntry:
    name: str
    size: int = 0
    is_dir: bool = False
    mode: int = 0


class ToolboxClient:
    """Client for the in-guest toolbox agent.

    Bound to a single UDS path (the sandbox's host-visible socket).
    Every call dials a fresh UDS → no long-lived connection, no
    keep-alive — matches the agent's Connection: close response
    shape and keeps the client simple.
    """

    def __init__(
        self,
        uds_path: str,
        port: int = TOOLBOX_VSOCK_PORT,
        dial_timeout: float = 5.0,
    ):
        self.uds_path = uds_path
        self.port = port
        self.dial_timeout = dial_timeout

    # ---- Health ----

    def health(self) -> Dict[str, object]:
        """GET /healthz on the agent. Returns the JSON body.
        Raises ToolboxError on non-200."""
        status, _, body = self._request("GET", "/healthz")
        if status != 200:
            raise ToolboxError(f"health: status={status} body={body[:200]!r}")
        return json.loads(body)

    # ---- Exec ----

    def exec(
        self,
        cmd: List[str],
        env: Optional[List[str]] = None,
        workdir: str = "",
        stdin: Optional[bytes] = None,
        timeout: float = 30.0,
    ) -> ExecResult:
        """Run a command in the guest and collect stdout/stderr/exit.
        Blocking — for streaming variant see `exec_stream`."""
        stdout = io.BytesIO()
        stderr = io.BytesIO()
        exit_code = -1
        for channel, payload in self.exec_stream(cmd, env=env, workdir=workdir, stdin=stdin, timeout=timeout):
            if channel == CHANNEL_STDOUT:
                stdout.write(payload)
            elif channel == CHANNEL_STDERR:
                stderr.write(payload)
            elif channel == CHANNEL_EXIT:
                exit_code = _parse_exit(payload)
        return ExecResult(exit_code=exit_code, stdout=stdout.getvalue(), stderr=stderr.getvalue())

    def exec_stream(
        self,
        cmd: List[str],
        env: Optional[List[str]] = None,
        workdir: str = "",
        stdin: Optional[bytes] = None,
        timeout: float = 30.0,
    ) -> Iterator[Tuple[int, bytes]]:
        """Yield (channel, payload) frames as they arrive. Channel
        constants: CHANNEL_STDOUT, CHANNEL_STDERR, CHANNEL_EXIT.

        Called by exec() to aggregate; call directly for line-by-line
        streaming (see examples/python/cookbook/exec_stream.py).
        """
        body: Dict[str, object] = {"cmd": cmd}
        if env:
            body["env"] = env
        if workdir:
            body["workdir"] = workdir

        sock = self._dial_connect()
        sock.settimeout(timeout)
        try:
            # Send HTTP request + body.
            body_bytes = json.dumps(body).encode("utf-8")
            http_req = (
                f"POST /exec HTTP/1.0\r\n"
                f"Host: x\r\n"
                f"Content-Length: {len(body_bytes)}\r\n"
                f"Content-Type: application/json\r\n"
                f"Connection: close\r\n"
                f"\r\n"
            ).encode("ascii")
            sock.sendall(http_req)
            sock.sendall(body_bytes)
            if stdin:
                # Frame the stdin as channel 0 so the agent sees
                # it, then signal EOF via a zero-length stdin frame.
                _write_frame(sock, CHANNEL_STDIN, stdin)
                _write_frame(sock, CHANNEL_STDIN, b"")

            # Read HTTP response line + headers (drain until blank line).
            f = sock.makefile("rb")
            status_line = f.readline()
            if not status_line.startswith(b"HTTP/"):
                raise ToolboxError(f"exec: unexpected response: {status_line!r}")
            while True:
                line = f.readline()
                if line in (b"\r\n", b"\n", b""):
                    break

            # Stream framed responses.
            while True:
                hdr = f.read(5)
                if len(hdr) < 5:
                    return
                channel = hdr[0]
                n = struct.unpack(">I", hdr[1:5])[0]
                if n > 0:
                    payload = f.read(n)
                    if len(payload) < n:
                        return
                else:
                    payload = b""
                yield channel, payload
                if channel == CHANNEL_EXIT:
                    return
        finally:
            try:
                sock.close()
            except Exception:
                pass

    # ---- Files (basic list/upload/download) ----

    def list_files(self, path: str) -> List[FileEntry]:
        status, _, body = self._request("GET", f"/files?path={path}")
        if status != 200:
            raise ToolboxError(f"list_files: status={status} body={body[:200]!r}")
        parsed = json.loads(body)
        return [
            FileEntry(
                name=e.get("name", ""),
                size=e.get("size", 0),
                # Agent returns kind="file"|"dir"; map to is_dir bool.
                is_dir=e.get("kind") == "dir",
                mode=e.get("mode", 0),
            )
            for e in parsed.get("entries", [])
        ]

    def download(self, path: str) -> bytes:
        status, _, body = self._request("GET", f"/files/download?path={path}", raw_body=True)
        if status != 200:
            raise ToolboxError(f"download: status={status} body={body[:200]!r}")
        return body if isinstance(body, bytes) else body.encode("utf-8")

    def upload(self, path: str, data: bytes) -> None:
        status, _, body = self._request(
            "POST",
            f"/files/upload?path={path}",
            body=data,
            content_type="application/octet-stream",
        )
        if status not in (200, 201):
            raise ToolboxError(f"upload: status={status} body={body[:200]!r}")

    def delete_file(self, path: str) -> None:
        status, _, body = self._request("DELETE", f"/files?path={path}")
        if status not in (200, 204):
            raise ToolboxError(f"delete_file: status={status} body={body[:200]!r}")

    def mkdir(self, path: str, parents: bool = False) -> None:
        # Agent expects {"path": ..., "all": bool} — not "parents".
        body = {"path": path, "all": parents}
        status, _, resp = self._request("POST", "/files/mkdir", body=json.dumps(body).encode())
        if status != 200:
            raise ToolboxError(f"mkdir: status={status} body={resp[:200]!r}")

    def rename(self, src: str, dst: str) -> None:
        # Agent expects {"old_path": ..., "new_path": ...}.
        body = {"old_path": src, "new_path": dst}
        status, _, resp = self._request("POST", "/files/rename", body=json.dumps(body).encode())
        if status != 200:
            raise ToolboxError(f"rename: status={status} body={resp[:200]!r}")

    def chmod(self, path: str, mode: int) -> None:
        body = {"path": path, "mode": mode}
        status, _, resp = self._request("POST", "/files/chmod", body=json.dumps(body).encode())
        if status != 200:
            raise ToolboxError(f"chmod: status={status} body={resp[:200]!r}")

    # ---- Git ----

    def git_clone(self, repository: str, directory: str, ref: str = "") -> Dict[str, object]:
        body: Dict[str, object] = {"repository": repository, "directory": directory}
        if ref:
            body["ref"] = ref
        status, _, resp = self._request("POST", "/git/clone", body=json.dumps(body).encode())
        if status != 200:
            raise ToolboxError(f"git_clone: status={status} body={resp[:200]!r}")
        return json.loads(resp)

    def git_status(self, directory: str) -> Dict[str, object]:
        body = {"directory": directory}
        status, _, resp = self._request("POST", "/git/status", body=json.dumps(body).encode())
        if status != 200:
            raise ToolboxError(f"git_status: status={status} body={resp[:200]!r}")
        return json.loads(resp)

    # ---- Secrets ----

    def set_secret(self, name: str, value: str) -> None:
        body = {"name": name, "value": value}
        status, _, resp = self._request("POST", "/secrets", body=json.dumps(body).encode())
        if status not in (200, 201):
            raise ToolboxError(f"set_secret: status={status} body={resp[:200]!r}")

    def list_secrets(self) -> List[str]:
        status, _, resp = self._request("GET", "/secrets")
        if status != 200:
            raise ToolboxError(f"list_secrets: status={status} body={resp[:200]!r}")
        parsed = json.loads(resp)
        # Agent returns {"secrets": [name1, name2, ...]}.
        return parsed.get("secrets", [])

    def delete_secret(self, name: str) -> None:
        status, _, resp = self._request("DELETE", f"/secrets/{name}")
        if status not in (200, 204):
            raise ToolboxError(f"delete_secret: status={status} body={resp[:200]!r}")

    # ---- Internals ----

    def _dial_connect(self) -> socket.socket:
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.settimeout(self.dial_timeout)
        try:
            s.connect(self.uds_path)
        except socket.timeout:
            raise ToolboxTimeout(f"dial timeout: {self.uds_path}")
        except OSError as e:
            s.close()
            raise ToolboxError(f"dial {self.uds_path}: {e}")
        # CONNECT handshake.
        s.sendall(f"CONNECT {self.port}\n".encode("ascii"))
        f = s.makefile("rb")
        line = f.readline()
        if not line.startswith(b"OK"):
            s.close()
            raise ToolboxError(f"CONNECT rejected: {line.strip()!r}")
        # dial_timeout only bounds the handshake; long-running operations
        # (git clone, large uploads) set their own deadline via the
        # per-call timeout or exec_stream. Leaving a 5s read timeout here
        # truncates any request that waits longer than 5s for its
        # response body.
        s.settimeout(None)
        return s

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[bytes] = None,
        content_type: str = "application/json",
        raw_body: bool = False,
    ) -> Tuple[int, Dict[str, str], object]:
        sock = self._dial_connect()
        try:
            hdrs = f"{method} {path} HTTP/1.0\r\nHost: x\r\nConnection: close\r\n"
            if body:
                hdrs += f"Content-Length: {len(body)}\r\nContent-Type: {content_type}\r\n"
            hdrs += "\r\n"
            sock.sendall(hdrs.encode("ascii"))
            if body:
                sock.sendall(body)
            f = sock.makefile("rb")
            status_line = f.readline()
            if not status_line.startswith(b"HTTP/"):
                raise ToolboxError(f"unexpected response: {status_line!r}")
            parts = status_line.split(b" ", 2)
            status = int(parts[1])
            headers: Dict[str, str] = {}
            while True:
                line = f.readline()
                if line in (b"\r\n", b"\n", b""):
                    break
                if b":" in line:
                    k, v = line.split(b":", 1)
                    headers[k.strip().decode("ascii").lower()] = v.strip().decode("latin-1")
            # Bound the read by Content-Length if present. Without
            # this, f.read() waits for EOF — which never arrives
            # promptly over the UDS bridge because the agent's
            # Connection: close + flush races with the host-side
            # io.Copy. A stale socket timeout masqueraded as a
            # "hung mkdir" bug.
            body_bytes: bytes
            # Responses that definitionally have no body (RFC 7230).
            if status in (204, 205, 304):
                body_bytes = b""
            else:
                cl = headers.get("content-length", "")
                if cl:
                    try:
                        n = int(cl)
                        body_bytes = f.read(n) if n > 0 else b""
                    except ValueError:
                        body_bytes = f.read()
                else:
                    body_bytes = f.read()
            if raw_body:
                return status, headers, body_bytes
            return status, headers, body_bytes.decode("utf-8", errors="replace")
        finally:
            try:
                sock.close()
            except Exception:
                pass


# ---- Helpers ----


def _write_frame(sock: socket.socket, channel: int, payload: bytes) -> None:
    sock.sendall(bytes([channel]) + struct.pack(">I", len(payload)) + payload)


def _parse_exit(payload: bytes) -> int:
    """Exit frames carry a signed 32-bit integer (BE). The agent uses
    -1 for "killed by signal"; 128+|signum| is NOT encoded for us
    because the CLI doesn't need the distinction."""
    if len(payload) < 4:
        return -1
    return struct.unpack(">i", payload[:4])[0]
